pkg/addon/addon.go

```diff
@@ -3,129 +3,97 @@ package addon
-	objects := []runtime.Object{}
-	// if the installation namespace is not set, to keep consistent with addon-framework,
-	// using open-cluster-management-agent-addon namespace as default namespace.
-	installNamespace := addon.Spec.InstallNamespace
-	if len(installNamespace) == 0 {
-		installNamespace = f.installNamespace
-	}
-
-	manifestConfig := struct {
-		KubeConfigSecret string
-		ClusterName      string
-		Namespace        string
-		Image            string
-	}{
-		KubeConfigSecret: fmt.Sprintf("%s-hub-kubeconfig", f.GetAgentAddonOptions().AddonName),
-		Namespace:        installNamespace,
-		ClusterName:      cluster.Name,
-		Image:            f.imageName,
-	}
+type Values struct {
+	IsOCP        bool         `json:"isOCP,omitempty"`
+	GlobalValues GlobalValues `json:"global,omitempty,omitempty"`
+}
 
-	for _, file := range agentDeploymentFiles {
-		template, err := manifestFiles.ReadFile(file)
-		if err != nil {
-			return objects, err
+func NewGetValuesFunc(imageName string) addonfactory.GetValuesFunc {
+	return func(cluster *clusterv1.ManagedCluster,
+		addon *addonapiv1alpha1.ManagedClusterAddOn) (addonfactory.Values, error) {
+		addonValues := Values{
+			GlobalValues: GlobalValues{
+				ImagePullPolicy: corev1.PullIfNotPresent,
+				ImagePullSecret: "open-cluster-management-image-pull-credentials",
+				ImageOverrides: map[string]string{
+					"multicloud_manager": imageName,
+				},
+				NodeSelector: map[string]string{},
+			},
 		}
-		raw := assets.MustCreateAssetFromTemplate(file, template, &manifestConfig).Data
-		object, _, err := genericCodec.Decode(raw, nil, nil)
+		for _, claim := range cluster.Status.ClusterClaims {
+			if claim.Name == "product.open-cluster-management.io" && claim.Value == "OpenShift" {
+				addonValues.IsOCP = true
+			}
+		}
+		values, err := addonfactory.JsonStructToValues(addonValues)
 		if err != nil {
 			return nil, err
 		}
-		objects = append(objects, object)
+		return values, nil
 	}
-	return objects, nil
 }
 
-func (f *foundationAgent) GetAgentAddonOptions() agent.AgentAddonOptions {
-	return agent.AgentAddonOptions{
-		AddonName: f.addonName,
-		Registration: &agent.RegistrationOption{
-			CSRConfigurations: agent.KubeClientSignerConfigurations(f.addonName, f.addonName),
-			CSRApproveCheck: func(
-				cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn, csr *certificatesv1.CertificateSigningRequest) bool {
-				// TODO add more csr check here.
-				return true
-			},
-			PermissionConfig: func(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn) error {
-				return f.createOrUpdateRoleBinding(cluster.Name)
-			},
+func NewRegistrationOption(kubeClient kubernetes.Interface, addonName string) *agent.RegistrationOption {
+	return &agent.RegistrationOption{
+		CSRConfigurations: agent.KubeClientSignerConfigurations(addonName, addonName),
+		CSRApproveCheck: func(
```

Comment:
i think you can use the approve helper in addon-framework

Reply:
done